### ISSUE TITLE
Добавление конфигурационных параметров для настройки тестов

### ISSUE THREAD

**Issue by artbear** [xDrivenDevelopment/xUnitFor1C#513]
Есть тесты, которые являются универсальными для многих конфигураций.
Нужна возможность их настраивать, чтобы разделить реализацию тесты и настройку под конкретную конфигурацию.

Например, для тестов открытия форм конфигурации нужно задавать исключения форм, которые не требуется тестировать.
Для тестов пользователей с ограниченными правами нужно задавать список пользователей и/или наборы прав, с которыми нужно запускать эти тесты.

Предлагаю следующее:
- при тестировании возможно указать настроечный xml-файл, в котором будут храниться индивидуальные настройки для тестов
  - при интерактивном тестировании файл можно выбрать вручную
  - при запуске из командной строки будет добавлен спец.параметр `config-file`, указывающий на путь к файлу настройки
  - формат хмл-файла - это простое дерево, где узлом является имя/идентификатор тестового набора, а листьями являются конкретные параметры для тестирования
- в ядро xUnitFor1С (для сервера и УФ-клиента) будет добавлен набор методов для получения параметров настройки

Обсудим?

**Comment by pumbaEO** [comment]
Ты в юнит тесты пытаешься засунунть интеграционные тесты. 

Простая генерилка текста bdd позволяет тебе тебе уже получить разлчиные настройки для сценариев. В частности у меня так выглядит тестирование открытия форм для различных подсистем. 
![cucumber1](https://cloud.githubusercontent.com/assets/166003/8067945/a56ccca0-0ef9-11e5-952e-de12e85f7650.png) 

и код тестовой обработки вот так выглядит
![cucumber2](https://cloud.githubusercontent.com/assets/166003/8068045/21d2ed7e-0efa-11e5-8a0c-2a01a0c96253.png)

**Comment by artbear** [comment]
@pumbaEO Женя, я уже давно не говорю "юнит-тесты", а в нашем проекте работаю над различными видами тестов
Сам знаешь, в рамках 1С очень сложно сделать "чистые" юнит-тесты :)

И настройка тестов под конкретную конфигурацию ИМХО не является злом :)

подход с БДД очень интересен, и мы к нему движемся.
Картинки у тебя интересные.

Я фактически и подготовил релиз 3.0 как релиз, позволяющий нашему проекту работать со  сценарными тестами и БДД.

В БДД есть очень полезные фишки, например, повторное использование шагов мне очень нравится. Но пока не все проблемы использования решены :(

Например, пока что мне очень не нравится то, что нужно юзать некие генераторы текста/кода, это замедляет процесс разработки/тестирования. Наверняка, со мной ты и кто-то еще не согласятся, но это мое текущее видение.

Женя, опиши, пожалуйста, в чем сейчас плюсы и минусы твоей доработки для Огурца.
Напиши свое видение, как можно юзать эту доработку.
Жду, очень интересно.

**Comment by pumbaEO** [comment]
1. Работает в linux и windows. 
2. Тест обратно совместим с юнит тестами, т.е. когда идет генерация сценария то формирую только процедуру "ПолучитьСписокТестов" и туда в массив добавляю все шаги (хоть они и могут повторятся). 
3. Повторно используемый код, т.е. приоритет поиска процедуры для выполнения ищу по текущей фиче, если не нахожу, то ищу по всему дереву тестов. Позволяет мне вынести библиотечные функции и тесты в отдельные обработки и использовать их повторно.

**Comment by artbear** [comment]
Это плюсы.
А замеченные недостатки есть?

**Comment by pumbaEO** [comment]
Пока одно замечание скорость, за счет парсинга и таймаутов время теряется на обмен данными, а не на тесты.

**Comment by artbear** [comment]
@pumbaEO по БДД понятно.

А что еще ты можешь сказать по конфигурационной настройке?
Нужно вообще? 
если нет, как реализовать указанную доп.настройку тестов (без БДД) ?
если да, что скажешь по формату настройки? может быть, предложишь другие варианты?

Кто еще что может добавить?
@allustin ? @kuntashov ? @wizi4d ? @ValeraS ?

**Comment by ValeraS** [comment]
Мне кажется это функциональность самого теста а не xUnitFor1С. Что мешает тест-обработке при заполнении списка тестовых методов прочитать свой конфигурационный файл и выполнить необходимые настройки?

**Comment by artbear** [comment]
Согласен, но в режиме УФ на сервере у формы нет доступа к файлу на клиенте, т.е. тест тупо свои настройки прочитать не сможет :(

Правда, это можно сделать в `ПолучитьСписокТестов` при его вызове с клиента.
Но чисто серверный тест, который определен только в модуле обработки, даже при чтении внутри  `ПолучитьСписокТестов` файл с клиента не получит, т.к. этот метод будет вызван только на сервере :(

**Comment by wizi4d** [comment]
@ValeraS +1
Иначе получится сильная связанность, ядро будет знать детали, о которых знать не должно.

Можно пример серверного теста реально требующего конфигурирование в отдельном файле?

**Comment by EvilBeaver** [comment]
Нужен конкретный use-case - для чего будет использоваться данный функционал. Просто, имхо, проблем больше, чем получаемой выгоды

**Comment by EvilBeaver** [comment]
@wizi4d - опередил

**Comment by artbear** [comment]
Злые вы :)
В процессе обсуждения с Андреем (@EvilBeaver ) я уже понял, что оба описанных мной сценария выполняются на клиенте и там нет проблемы с доступом к файлу.

**Comment by artbear** [comment]
>Мне кажется это функциональность самого теста а не xUnitFor1С. Что мешает тест-обработке, при заполнении списка тестовых методов, прочитать свой конфигурационный файл и выполнить необходимые настройки?

Поясню свои размышления при создании задачи - функционал (чтение настроек) может юзаться в нескольких сценариях. А известно, что если функционал пересекается в нескольких местах, почти всегда лучше его собрать в одном месте (например, ядре или плагинах) и исключить дублирование.
Так и возникла эта задача

**Comment by artbear** [comment]
Вот за что я и люблю коллективную разработку - захочешь сделать одно, тебя или собьют с пути, направив на правильный :) , или дадут информацию к размышлению, или расскажут, что нужная фича уже давно реализована :)

**Comment by artbear** [comment]
или докажут, что новая фича вообще не нужна :)

**Comment by artbear** [comment]
Я же упорный :)
Поэтому набор следующих вопросов:
-  где должен находиться этот конфигурационный/настроечный файл?
  - Если это внешняя обработка, напрашивается мысль положить файл настройки рядом с обработкой-тестом. Но в этом случае опять идет смешение кода исходников и кода настройки, что мне не очень нравится.

  - А если это встроенная обработка, где должен находиться файл настройки?

**Comment by pumbaEO** [comment]
devops - все есть код, т.е. я пока не вижу проблем файла настроек рядом с обработкой. 
Если встроенная, тогда в макете. 

И опять таки, универсального кода не сделаешь, тест у тебя работает на том коде который сейчас в конфигурации, настройки тоже для определенного вида теста, почему тогда ты уверен, что настройки можно будет между разными конфигурациями тасовать?

**Comment by artbear** [comment]
Настройки индивидуальны для ИБ, а вот код теста независим от конфы и
хочется именно тест распространять как  продукт, релиз

пт, 12 Июн 2015, 9:38, Shenja Sosna <notifications@github.com>:
